### PR TITLE
[proxima-core] #170 implement ThroughputLimiter

### PR DESCRIPTION
**Pull request by je-ik**

### CODE REVIEW

File: direct/core/src/main/java/cz/o2/proxima/direct/commitlog/ThroughputLimitedCommitLogReader.java [hunk 1]
```diff
@@ -0,0 +1,114 @@
...
+   * @param delegate delegate to read data from
+   * @param limiter the throughput limiter
+   * @return throughput limited {@link CommitLogReader}
+   */
+  public static CommitLogReader withThroughputLimit(
+      CommitLogReader delegate, ThroughputLimiter limiter) {
+    return new LimitedCommitLogReader(delegate, limiter);
+  }
+
+  private static class DelegateCommitLogReader implements CommitLogReader {
```
Comment (dmvk) [inline direct/core/src/main/java/cz/o2/proxima/direct/commitlog/ThroughputLimitedCommitLogReader.java:33]: I'd follow naming conventions from google projects here (guava, grpc, ...) -> `ForwardingCommitLogReader`

**Comment by je-ik** [comment]
@dmvk this is ready for review. I'm still tuning the code coverage (and hence I have to touch sort of unrelated pieces of code - e.g. cassandra). I will keep these commits separate and squash the rest after review.

**Comment by je-ik** [comment]
@dmvk checks passed

**Comment by je-ik** [comment]
This closes O2-Czech-Republic#170

File: core/src/main/java/cz/o2/proxima/storage/ThroughputLimiter.java [hunk 2]
```diff
@@ -0,0 +1,72 @@
...
+    public Duration getPauseTime(Context context) {
+      return Duration.ZERO;
+    }
+  }
+
+  /** A context of {@link #getPauseTime} method. */
+  interface Context {
+
+    /**
+     * Get collection of {@link Partition Partitions} associated with reaeder reading with this
```
Comment (dmvk) [inline core/src/main/java/cz/o2/proxima/storage/ThroughputLimiter.java:45]: typo: reader

File: core/src/main/java/cz/o2/proxima/storage/ThroughputLimiter.java [hunk 3]
```diff
@@ -0,0 +1,72 @@
...
+     * limiter
+     *
+     * @return list of associated partitions
+     */
+    Collection<Partition> getConsumedPartitions();
+
+    /** Get count of all partitions in the reader. */
+    int getNumPartitions();
```
Comment (dmvk) [inline core/src/main/java/cz/o2/proxima/storage/ThroughputLimiter.java:53]: is internal implementation doing anything else than `getConsumedPartitions().size()`? do we need this method?

File: core/src/main/java/cz/o2/proxima/storage/ThroughputLimiter.java [hunk 4]
```diff
@@ -0,0 +1,72 @@
...
+  /** A context of {@link #getPauseTime} method. */
+  interface Context {
+
+    /**
+     * Get collection of {@link Partition Partitions} associated with reaeder reading with this
+     * limiter
+     *
+     * @return list of associated partitions
+     */
+    Collection<Partition> getConsumedPartitions();
```
Comment (dmvk) [inline core/src/main/java/cz/o2/proxima/storage/ThroughputLimiter.java:50]: -> `getAssociatedPartions` or just `getPartitions` seems better

File: core/src/main/java/cz/o2/proxima/storage/ThroughputLimiter.java [hunk 5]
```diff
@@ -0,0 +1,72 @@
...
+
+import cz.o2.proxima.annotations.Evolving;
+import cz.o2.proxima.annotations.Internal;
+import java.io.Serializable;
+import java.time.Duration;
+import java.util.Collection;
+
+/** A limiter of data rate coming from various sources. */
+@Evolving
+public interface ThroughputLimiter extends Serializable {
```
Comment (dmvk) [inline core/src/main/java/cz/o2/proxima/storage/ThroughputLimiter.java:26]: Do we really need ThroughputLimiter to be serializable? We should be always able to re-construct it from the config right?

File: direct/core/src/main/java/cz/o2/proxima/direct/batch/BatchLogObservers.java [hunk 6]
```diff
@@ -0,0 +1,65 @@
...
+import cz.o2.proxima.annotations.Internal;
+import cz.o2.proxima.direct.batch.BatchLogObserver.OnNextContext;
+import cz.o2.proxima.storage.Partition;
+import cz.o2.proxima.time.WatermarkSupplier;
+
+/** Utility class related to {@link BatchLogObserver BatchLogObservers}. */
+@Internal
+public class BatchLogObservers {
+
+  public static OnNextContext defaultContext(Partition partition) {
```
Comment (dmvk) [inline direct/core/src/main/java/cz/o2/proxima/direct/batch/BatchLogObservers.java:27]: This method deserves javadoc ;) other methods seem intuitive, so no need there

File: direct/core/src/main/java/cz/o2/proxima/direct/batch/BatchLogReaders.java [hunk 7]
```diff
@@ -0,0 +1,184 @@
...
+        BatchLogObserver observer) {
+
+      super.observe(partitions, attributes, throughputLimited(observer, partitions));
+    }
+
+    @Override
+    public String toString() {
+      return MoreObjects.toStringHelper(this)
+          .add("limiter", limiter)
+          .add("delegate", super.toString())
```
Comment (dmvk) [inline direct/core/src/main/java/cz/o2/proxima/direct/batch/BatchLogReaders.java:83]: nit: `delegate.toString()`

File: direct/core/src/main/java/cz/o2/proxima/direct/batch/BatchLogReaders.java [hunk 8]
```diff
@@ -0,0 +1,184 @@
...
+
+      super(delegate);
+      this.numPartitions = numPartitions;
+      this.assignedPartitions = new ArrayList<>(assignedPartitions);
+      this.limiter = Objects.requireNonNull(limiter);
+    }
+
+    @Override
+    public boolean onNext(StreamElement element) {
+      if (throwsInterrupted(this::waitIfNecessary)) {
```
Comment (dmvk) [inline direct/core/src/main/java/cz/o2/proxima/direct/batch/BatchLogReaders.java:125]: Not sure about threading model here. If every reader has dedicated thread, sleeping here should be ok.

File: direct/core/src/main/java/cz/o2/proxima/direct/core/DirectDataOperator.java [hunk 9]
```diff
@@ -396,4 +397,86 @@ public void close() {
   public Repository getRepository() {
     return repo;
   }
+
+  @VisibleForTesting
+  public static class DelegateDataAccessorFactory implements DataAccessorFactory {
```
Comment (dmvk) [inline direct/core/src/main/java/cz/o2/proxima/direct/core/DirectDataOperator.java:402]: why do we need the wrapping class?

**Review by dmvk — APPROVED** [review]
LGTM, big thumbs up for this PR 👍  I left few questions and comments, but this looks good to merge in general.

cc @tisonet

File: direct/core/src/main/java/cz/o2/proxima/direct/commitlog/ThroughputLimitedCommitLogReader.java [hunk 10]
```diff
@@ -0,0 +1,128 @@
...
+
+    @Delegate private final CommitLogReader delegate;
+
+    private DelegateCommitLogReader(CommitLogReader delegate) {
+      this.delegate = delegate;
+    }
+  }
+
+  private static class LimitedCommitLogReader extends DelegateCommitLogReader {
```
[the remark this answers is not in the document]
Comment (je-ik) [reply direct/core/src/main/java/cz/o2/proxima/direct/commitlog/ThroughputLimitedCommitLogReader.java]: What's the reasoning about that?

File: core/src/main/java/cz/o2/proxima/storage/ThroughputLimiter.java [hunk 11]
```diff
@@ -0,0 +1,72 @@
...
+
+  @Internal
+  class NoOpThroughputLimiter implements ThroughputLimiter {
```
[the remark this answers is not in the document]
Comment (je-ik) [reply core/src/main/java/cz/o2/proxima/storage/ThroughputLimiter.java:29]: This is actually unused. Probably will be removed completely.

File: core/src/main/java/cz/o2/proxima/storage/ThroughputLimiter.java (hunk 4 above)
Comment (je-ik) [reply core/src/main/java/cz/o2/proxima/storage/ThroughputLimiter.java:50]: `getPartitions` is ambiguous (is it all partitions, or just a consumed subset).
What about `getObservedPartitions()`? After all, it is result of `reader.observe()`, tak partitions are se to the reader.

**Comment by je-ik** [comment]
This closes O2-Czech-Republic#170

**Pull request merged**